From patchwork Mon Oct 16 15:31:38 2017
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Ulrich Hecht <ulrich.hecht+renesas@gmail.com>
X-Patchwork-Id: 10010975
Return-Path: <dri-devel-bounces@lists.freedesktop.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
	[172.30.200.125])
	by pdx-korg-patchwork.web.codeaurora.org (Postfix) with ESMTP id
	E4E6460211 for <patchwork-dri-devel@patchwork.kernel.org>;
	Tue, 17 Oct 2017 07:36:19 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id D5EC32871D
	for <patchwork-dri-devel@patchwork.kernel.org>;
	Tue, 17 Oct 2017 07:36:19 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id C95D12876A; Tue, 17 Oct 2017 07:36:19 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-4.1 required=2.0 tests=BAYES_00,
	DKIM_ADSP_CUSTOM_MED,
	DKIM_SIGNED, FREEMAIL_FROM, RCVD_IN_DNSWL_MED,
	T_DKIM_INVALID autolearn=ham version=3.3.1
Received: from gabe.freedesktop.org (gabe.freedesktop.org [131.252.210.177])
	(using TLSv1.2 with cipher DHE-RSA-AES256-GCM-SHA384 (256/256
	bits)) (No client certificate requested)
	by mail.wl.linuxfoundation.org (Postfix) with ESMTPS id 499582871D
	for <patchwork-dri-devel@patchwork.kernel.org>;
	Tue, 17 Oct 2017 07:36:19 +0000 (UTC)
Received: from gabe.freedesktop.org (localhost [127.0.0.1])
	by gabe.freedesktop.org (Postfix) with ESMTP id 5DD146E5BA;
	Tue, 17 Oct 2017 07:35:26 +0000 (UTC)
X-Original-To: dri-devel@lists.freedesktop.org
Delivered-To: dri-devel@lists.freedesktop.org
Received: from mail-wm0-x244.google.com (mail-wm0-x244.google.com
	[IPv6:2a00:1450:400c:c09::244])
	by gabe.freedesktop.org (Postfix) with ESMTPS id B05B86E48A
	for <dri-devel@lists.freedesktop.org>;
	Mon, 16 Oct 2017 15:32:01 +0000 (UTC)
Received: by mail-wm0-x244.google.com with SMTP id b189so4513083wmd.4
	for <dri-devel@lists.freedesktop.org>;
	Mon, 16 Oct 2017 08:32:01 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=gmail.com; s=20161025;
	h=sender:from:to:cc:subject:date:message-id:in-reply-to:references;
	bh=qTT3fcSslWmq2/xfCt7UHhMau7eqzrbwyj/v2dChwDM=;
	b=ebADTWnOWN15noj6Es4s9os/IQFeqBbClNSD3NeI9POK3qr+dfyGGGT3Ar22gjU+sD
	pK66QCz3QGgjceZZfLBdr1g7kYek6O4Qbhz1pGbpohlQOYuORxApmxTrcl7T415wKZVK
	gTO/qjfoUJLeVNy+hB7kQ2MfFLFpZSIdoC6Ud90QQ9egv1m7Q7Fj6HekWE6V7dP2KGZg
	tr94qilxo5Xn0WF9iNAB62WaD0YQ1zlzpqm3h8/dj4qScv7l2TaW9MKUUvfOKIZhvj9U
	LsvZ/poCcb6yMTgbY1LACPcQvrVuHH6H+YIY5JdSJgV7VESVt24jyqvLqGPSiZYmTXdl
	ea4w==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
	d=1e100.net; s=20161025;
	h=x-gm-message-state:sender:from:to:cc:subject:date:message-id
	:in-reply-to:references;
	bh=qTT3fcSslWmq2/xfCt7UHhMau7eqzrbwyj/v2dChwDM=;
	b=MQZWlFo1AmepMq50KcHmKGa4aBMA/asL3U1dkgulRNQ8owRHfM6quU3rnkual2MRAv
	kw40VEAcz1uUEVK/RRwt1mL/y81FbRJhxuPErd+O43XDe4rrGMA+XR9z2PTY/xwbzYEN
	nev2OGIc4+BtRQ/4SvvHWPOhsqqaOGa+ICjxO4HXMg/l98CKFyFpzHTCkCeBzzKX2IZ1
	FbFNBA8t/zCrSpa5HV3+CryrAqSYnP2rIVtTZo/6W0ienaJ5cT0Iabvo8xeF1Jq7SDtN
	t4ZZU/r+b2QQnYzaW1rzKFfEtn2dCE0+ZbkPP6gNul412PzkMoz72gAk+532L7m+Q1fz
	55mQ==
X-Gm-Message-State: AMCzsaXvIfTAWDtq03qEBk56OZP696oYFQD+5ywqA3JwdRPyXIdRiqSw
	RWBpEZsUfPDek6l8Gi10DA==
X-Google-Smtp-Source: 
 ABhQp+QGaBxHgPZvVCw2Z0CoEUs6V4SPgIDxPpgGd6vW0BymcNxS+BcnENFABk1vzEKEjXqYx3UfMQ==
X-Received: by 10.28.5.148 with SMTP id 142mr1209023wmf.142.1508167920314;
	Mon, 16 Oct 2017 08:32:00 -0700 (PDT)
Received: from groucho.site (ipbcc0ce42.dynamic.kabel-deutschland.de.
	[188.192.206.66]) by smtp.gmail.com with ESMTPSA id
	v2sm6897751wmf.40.2017.10.16.08.31.59
	(version=TLS1_2 cipher=ECDHE-RSA-AES128-SHA bits=128/128);
	Mon, 16 Oct 2017 08:31:59 -0700 (PDT)
From: Ulrich Hecht <ulrich.hecht+renesas@gmail.com>
To: linux-mediatek@lists.infradead.org
Subject: [RFC v2 6/8] hack: mediatek: get mmsys to register as both DRM and
	clock device
Date: Mon, 16 Oct 2017 17:31:38 +0200
Message-Id: <1508167900-27415-7-git-send-email-ulrich.hecht+renesas@gmail.com>
X-Mailer: git-send-email 2.7.4
In-Reply-To: 
 <1508167900-27415-1-git-send-email-ulrich.hecht+renesas@gmail.com>
References: <1508167900-27415-1-git-send-email-ulrich.hecht+renesas@gmail.com>
X-Mailman-Approved-At: Tue, 17 Oct 2017 07:35:19 +0000
Cc: jacopo@jmondi.org, magnus.damm@gmail.com, dri-devel@lists.freedesktop.org,
	Ulrich Hecht <ulrich.hecht+renesas@gmail.com>,
	laurent.pinchart@ideasonboard.com, robin.murphy@arm.com,
	linux-arm-kernel@lists.infradead.org
X-BeenThere: dri-devel@lists.freedesktop.org
X-Mailman-Version: 2.1.18
Precedence: list
List-Id: Direct Rendering Infrastructure - Development
	<dri-devel.lists.freedesktop.org>
List-Unsubscribe: <https://lists.freedesktop.org/mailman/options/dri-devel>,
	<mailto:dri-devel-request@lists.freedesktop.org?subject=unsubscribe>
List-Archive: <https://lists.freedesktop.org/archives/dri-devel>
List-Post: <mailto:dri-devel@lists.freedesktop.org>
List-Help: <mailto:dri-devel-request@lists.freedesktop.org?subject=help>
List-Subscribe: <https://lists.freedesktop.org/mailman/listinfo/dri-devel>,
	<mailto:dri-devel-request@lists.freedesktop.org?subject=subscribe>
MIME-Version: 1.0
Errors-To: dri-devel-bounces@lists.freedesktop.org
Sender: "dri-devel" <dri-devel-bounces@lists.freedesktop.org>
X-Virus-Scanned: ClamAV using ClamSMTP

"mediatek,mt8173-mmsys" is the compatible string for both a clock controller
and a DRM device. The assumption seems to be that all compatible strings
trigger, and that mmsys will be registered as two devices.
This does work in the 3.18 vendor kernel, but it is not what happens in
mainline: only the clock portion is probed.

This hack probes the clocks from mtk_drm_probe(), and has all drivers
that need these clocks defer instead of fail. Fixes DRM initialization
on Acer Chromebook R13.

Signed-off-by: Ulrich Hecht <ulrich.hecht+renesas@gmail.com>
---
 drivers/clk/mediatek/clk-mt8173.c      | 4 ++--
 drivers/gpu/drm/mediatek/mtk_dpi.c     | 5 +++--
 drivers/gpu/drm/mediatek/mtk_drm_ddp.c | 5 +++--
 drivers/gpu/drm/mediatek/mtk_drm_drv.c | 7 +++++++
 drivers/gpu/drm/mediatek/mtk_hdmi.c    | 5 +++--
 5 files changed, 18 insertions(+), 8 deletions(-)

diff --git a/drivers/clk/mediatek/clk-mt8173.c b/drivers/clk/mediatek/clk-mt8173.c
index 96c292c..9339d2c 100644
--- a/drivers/clk/mediatek/clk-mt8173.c
+++ b/drivers/clk/mediatek/clk-mt8173.c
@@ -1152,7 +1152,7 @@ static void __init mtk_imgsys_init(struct device_node *node)
 }
 CLK_OF_DECLARE(mtk_imgsys, "mediatek,mt8173-imgsys", mtk_imgsys_init);
 
-static void __init mtk_mmsys_init(struct device_node *node)
+/*static*/ void /*__init*/ mtk_mmsys_init(struct device_node *node)
 {
 	struct clk_onecell_data *clk_data;
 	int r;
@@ -1167,7 +1167,7 @@ static void __init mtk_mmsys_init(struct device_node *node)
 		pr_err("%s(): could not register clock provider: %d\n",
 			__func__, r);
 }
-CLK_OF_DECLARE(mtk_mmsys, "mediatek,mt8173-mmsys", mtk_mmsys_init);
+//CLK_OF_DECLARE(mtk_mmsys, "mediatek,mt8173-mmsys", mtk_mmsys_init);
 
 static void __init mtk_vdecsys_init(struct device_node *node)
 {
diff --git a/drivers/gpu/drm/mediatek/mtk_dpi.c b/drivers/gpu/drm/mediatek/mtk_dpi.c
index e80a603..ed999891 100644
--- a/drivers/gpu/drm/mediatek/mtk_dpi.c
+++ b/drivers/gpu/drm/mediatek/mtk_dpi.c
@@ -682,8 +682,9 @@ static int mtk_dpi_probe(struct platform_device *pdev)
 	dpi->engine_clk = devm_clk_get(dev, "engine");
 	if (IS_ERR(dpi->engine_clk)) {
 		ret = PTR_ERR(dpi->engine_clk);
-		dev_err(dev, "Failed to get engine clock: %d\n", ret);
-		return ret;
+		dev_err(dev, "Failed to get engine clock: %d, deferring\n", ret);
+		return -EPROBE_DEFER;
+		//return ret;
 	}
 
 	dpi->pixel_clk = devm_clk_get(dev, "pixel");
diff --git a/drivers/gpu/drm/mediatek/mtk_drm_ddp.c b/drivers/gpu/drm/mediatek/mtk_drm_ddp.c
index 8130f3d..465ee60 100644
--- a/drivers/gpu/drm/mediatek/mtk_drm_ddp.c
+++ b/drivers/gpu/drm/mediatek/mtk_drm_ddp.c
@@ -382,8 +382,9 @@ static int mtk_ddp_probe(struct platform_device *pdev)
 
 	ddp->clk = devm_clk_get(dev, NULL);
 	if (IS_ERR(ddp->clk)) {
-		dev_err(dev, "Failed to get clock\n");
-		return PTR_ERR(ddp->clk);
+		dev_err(dev, "Failed to get clock, deferring\n");
+		return -EPROBE_DEFER;
+		//return PTR_ERR(ddp->clk);
 	}
 
 	regs = platform_get_resource(pdev, IORESOURCE_MEM, 0);
diff --git a/drivers/gpu/drm/mediatek/mtk_drm_drv.c b/drivers/gpu/drm/mediatek/mtk_drm_drv.c
index a2ca90f..ce99921 100644
--- a/drivers/gpu/drm/mediatek/mtk_drm_drv.c
+++ b/drivers/gpu/drm/mediatek/mtk_drm_drv.c
@@ -381,6 +381,8 @@ static const struct of_device_id mtk_ddp_comp_dt_ids[] = {
 	{ }
 };
 
+static int hack_mmsys_inited;
+void mtk_mmsys_init(struct device_node *node);
 static int mtk_drm_probe(struct platform_device *pdev)
 {
 	struct device *dev = &pdev->dev;
@@ -391,6 +393,11 @@ static int mtk_drm_probe(struct platform_device *pdev)
 	int ret;
 	int i;
 
+	if (!hack_mmsys_inited) {
+		mtk_mmsys_init(dev->of_node);
+		hack_mmsys_inited = 1;
+	}
+
 	private = devm_kzalloc(dev, sizeof(*private), GFP_KERNEL);
 	if (!private)
 		return -ENOMEM;
diff --git a/drivers/gpu/drm/mediatek/mtk_hdmi.c b/drivers/gpu/drm/mediatek/mtk_hdmi.c
index 690c675..715c0e0 100644
--- a/drivers/gpu/drm/mediatek/mtk_hdmi.c
+++ b/drivers/gpu/drm/mediatek/mtk_hdmi.c
@@ -1441,8 +1441,9 @@ static int mtk_hdmi_dt_parse_pdata(struct mtk_hdmi *hdmi,
 
 	ret = mtk_hdmi_get_all_clk(hdmi, np);
 	if (ret) {
-		dev_err(dev, "Failed to get clocks: %d\n", ret);
-		return ret;
+		dev_err(dev, "Failed to get clocks: %d, deferring\n", ret);
+		return -EPROBE_DEFER;
+		//return ret;
 	}
 
 	/* The CEC module handles HDMI hotplug detection */
